scipy/spatial/_hausdorff.pyx

```diff
@@ -16,56 +16,73 @@ import numpy as np
 cimport numpy as np
 cimport cython
 from libc.math cimport sqrt
+from libc.stdlib cimport srand, rand, RAND_MAX
+from libc.limits cimport UINT_MAX
 
 __all__ = ['directed_hausdorff']
 
+cdef extern from "numpy/npy_math.h":
+    double inf "NPY_INFINITY"
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision(True)
+cdef void index_shuffler(long [::1] arr, size_t size):
+    cdef:
+        size_t i = 0
+        size_t j
+        int t
+
+    while i < (size - 1):
+        j = i + rand() / (RAND_MAX / (size - i) + 1)
+        t = arr[j]
+        arr[j] = arr[i]
+        arr[i] = t
+        i += 1
+
 @cython.boundscheck(False)
-def directed_hausdorff(double[:,::1] ar1, double[:,::1] ar2, seed=0):
+def directed_hausdorff(const double[:,::1] ar1,
+                       const double[:,::1] ar2,
+                       seed=0):
 
     cdef double cmax, cmin, d
-    cdef bint no_break_occurred
     cdef int N1 = ar1.shape[0]
     cdef int N2 = ar2.shape[0]
     cdef int data_dims = ar1.shape[1]
     cdef unsigned int i, j, k
-    cdef unsigned int i_store = 0, j_store = 0, i_ret = 0, j_ret = 0
-    cdef long[:] resort1, resort2
+    cdef unsigned int i_ret = 0, j_ret = 0
+    cdef long[::1] resort1 = np.arange(N1)
+    cdef long[::1] resort2 = np.arange(N2)
 
     # shuffling the points in each array generally increases the likelihood of
     # an advantageous break in the inner search loop and never decreases the
     # performance of the algorithm
     rng = np.random.RandomState(seed)
-    resort1 = np.arange(N1)
-    resort2 = np.arange(N2)
-    rng.shuffle(resort1)
-    rng.shuffle(resort2)
-    ar1 = np.asarray(ar1)[resort1]
-    ar2 = np.asarray(ar2)[resort2]
+    srand(rng.randint(UINT_MAX, dtype=np.uint64))
+    index_shuffler(resort1, N1)
+    index_shuffler(resort2, N2)
 
     cmax = 0
     for i in range(N1):
-        no_break_occurred = True
-        cmin = np.inf
+        cmin = inf
         for j in range(N2):
-            d = 0
+            d = (ar1[resort1[i], 0] - ar2[resort2[j], 0])**2
```

Comment:
There does actually seem to be some benefit to replacing a reset assignment like `d = 0` with a "more useful" assignment to a value that will have to be assigned in eventually anyway. I might have guessed the compiler could just ignore the `d=0`, but maybe depends on flags.